src/EditorFeatures/Core/Implementation/ChangeSignature/AbstractChangeSignatureCommandHandler.cs

```diff
@@ -70,62 +71,86 @@ private bool ExecuteCommand(ITextView textView, ITextBuffer subjectBuffer, Comma
-                        document.Project.Solution);
-                }
+                return true;
+            }
+        }
 
-                if (finalSolution == null)
+        private static void HandleResult(ChangeSignatureResult result, Solution oldSolution, Workspace workspace, CommandExecutionContext context)
+        {
+            var notificationService = workspace.Services.GetRequiredService<INotificationService>();
+            if (!result.Succeeded)
+            {
+                if (result.ChangeSignatureFailureKind != null)
                 {
-                    // User clicked cancel.
-                    return true;
+                    ShowError(result.ChangeSignatureFailureKind.Value, context.OperationContext, notificationService);
                 }
 
-                using (var workspaceUndoTransaction = workspace.OpenGlobalUndoTransaction(FeaturesResources.Change_signature))
-                {
-                    if (!workspace.TryApplyChanges(finalSolution))
-                    {
-                        // TODO: handle failure
-                        return true;
-                    }
+                return;
+            }
 
-                    workspaceUndoTransaction.Commit();
-                }
+            if (result.ConfirmationMessage != null && !notificationService.ConfirmMessageBox(result.ConfirmationMessage, severity: NotificationSeverity.Warning))
+            {
+                return;
+            }
 
-                return true;
+            var finalSolution = result.UpdatedSolution;
+
+            var previewService = workspace.Services.GetService<IPreviewDialogService>();
+            if (previewService != null && result.PreviewChanges)
+            {
+                // We are about to show a modal UI dialog so we should take over the command execution
+                // wait context. That means the command system won't attempt to show its own wait dialog 
+                // and also will take it into consideration when measuring command handling duration.
+                context.OperationContext.TakeOwnership();
+                finalSolution = previewService.PreviewChanges(
+                    string.Format(EditorFeaturesResources.Preview_Changes_0, EditorFeaturesResources.Change_Signature),
+                    "vs.csharp.refactoring.preview",
+                    EditorFeaturesResources.Change_Signature_colon,
+                    result.Name,
+                    result.Glyph.GetValueOrDefault(),
+                    result.UpdatedSolution,
+                    oldSolution);
+            }
+
+            if (finalSolution == null)
+            {
+                // User clicked cancel.
+                return;
+            }
+
+            if (workspace.TryApplyChanges(finalSolution))
+            {
+                using var workspaceUndoTransaction = workspace.OpenGlobalUndoTransaction(FeaturesResources.Change_signature);
+                workspaceUndoTransaction.Commit();
```

Comment:
i literally don't know waht this is doing.  If i'm reading it right, it's saying:

try to update the solution.  If that succeeds, open a transaction and then immediately commit the empty transaction.

I have no idea what the purpose of the transaction is.  @jasonmalinowski how is that supposed to work?

Reply:
woops I think that's a bad merge.. will fix